python/cudaq/operator/cudm_state.py

```diff
@@ -0,0 +1,131 @@
+# ============================================================================ #
+# Copyright (c) 2022 - 2024 NVIDIA Corporation & Affiliates.                   #
+# All rights reserved.                                                         #
+#                                                                              #
+# This source code and the accompanying materials are made available under     #
+# the terms of the Apache License 2.0 which accompanies this distribution.     #
+# ============================================================================ #
+
+from cuquantum.densitymat import DenseMixedState, DensePureState, WorkStream
+import numpy, cupy, atexit
+from typing import Sequence
+from cupy.cuda.memory import MemoryPointer, UnownedMemory
+from ..mlir._mlir_libs._quakeDialects import cudaq_runtime
+
+
+# Wrap state data (on device memory) as a `cupy` array.
+# Note: the `cupy` array only holds a reference to the GPU memory buffer, no copy.
+def to_cupy_array(state):
+    tensor = state.getTensor()
+    pDevice = tensor.data()
+    dtype = cupy.complex128
+    sizeByte = tensor.get_num_elements() * tensor.get_element_size()
+    # Use `UnownedMemory` to wrap the device pointer
+    mem = UnownedMemory(pDevice, sizeByte, owner=state)
+    memptr = MemoryPointer(mem, 0)
+    cupy_array = cupy.ndarray(tensor.get_num_elements(),
+                              dtype=dtype,
+                              memptr=memptr)
+    return cupy_array
+
+
+# Helper to convert a state vector to a density matrix
+def ket2dm(ket: cupy.ndarray) -> cupy.ndarray:
+    return cupy.outer(ket.reshape((ket.size, 1)),
+                      cupy.conjugate(ket.reshape((ket.size, 1))))
+
+
+# Helper to create a 'coherent' state as a state vector.
+def coherent_state(N: int, alpha: float):
```

Comment:
I just want to clarify that it's not a complete clone of the referred code.
To prevent further confusion, I've refactored the above code to not use `cumprod` in implementing the analytical representation of the coherent state (https://github.com/NVIDIA/cuda-quantum/pull/2391/commits/7e4a7f27315a176649ad48ea3485640747e0a57a)